### ISSUE TITLE
Verbose log option

### ISSUE THREAD

**Issue by galfridus** [bafolts/terraforming-mars#1976]
Many triggered effects (and some main ones) do not show up in the log. For example, corporations that gain income or MC when certain tags are played. There's no indication that this happened other than watching someone's stats increase. This makes it easy to forget about such effects.

Similarly, there's no log indication for cards that increase income via a formula what the result of a formula was. Thus, a large increase can go unnoticed.

A "Verbose" log option which put all this stuff in the log would help visibility of game state (and particularly of changes to game state) a great deal. Ideally this would be always stored and a toggle, so that turning it on shows all past info.

**Comment by kevinb9n** [comment]
#1110 had forked into a few related topics. It started with the idea of a game log being downloadable, which led to the idea that that game log should be complete. Complete should mean you could write a program to calculate everyone's ending inventory etc. based on the log file alone.

As for what's shown in the UI, I would start by saying it should be complete, and there only needs to be a verbose toggle if anyone complains that it's too much information.

And maybe no one WILL even complain, if the information is presented in a nice way.  here was a discord conversation from last week.

![image](https://user-images.githubusercontent.com/934551/100557705-b5e68f00-325f-11eb-9193-7dec20f4ec3e.png)

**Comment by galfridus** [comment]
That looks really good!

**Comment by bafolts** [comment]
I think the approach of showing the delta is a great idea. The user can see what has changed and users who want to be able to see what the state of the game was at a certain point could rely on the delta and calculate the state versus requiring the entire state.

**Comment by kevinb9n** [comment]
I still want to reorganize this pile of interrelated issues and figure out which existing issue will serve as which.

* One issue about having a game log of any kind that is verified by test to be complete (the test can calculate the complete ending state of a game from its log and initial conditions alone, without knowing any game rules, just by watching pieces move around in this enormous log file)
* One issue about presenting that full information in the "game log" UI, *either* along with a "concise mode" button *or* by trying to present it compactly and attractively and hope everyone's just okay with it. (Both those options would need design work; how would we know exactly where "concise" has to draw the line?)

Actually maybe the either/or in the second bullet would each be their own bugs too.

**Comment by Wyko** [comment]
I would really love seeing an explanation for what actually happened in the Turmoil phase!

**Comment by kevinb9n** [comment]
I had some new thoughts on this.

I think the first most important thing to do is to include all the information no matter how ugly. Hit it with a blunt hammer: whenever the logInventoryChanges() (or whatever) function is called, take a snapshot of all the current amounts and calculate the deltas between those amounts and those from the previous snapshot. It's coarse, but it's also foolproof: nothing can accidentally be missed.

Roughly I think this should just be called once after every full action, corp/prelude play, research phase, and solar phase. But exactly when to do it is just a detail.

I think this would make bug reports more clear and reliable (not to mention that it would help bugs to be *noticed* more). I also think that if you can store all of these log files, searching through them may be an even more reliable measure of whether a bug was fixed. It's hard to come up with all the possible edge cases in unit tests, but real players in real games are great at it.

**Comment by nwai90** [comment]
> I would really love seeing an explanation for what actually happened in the Turmoil phase!

Agree with including Turmoil global event resolution. The fine-grained details of resource quantity changes should not be displayed by default imo. These are useful for debugging but add a lot of visual clutter which can be hidden behind a toggle.

<img width="391" alt="Screenshot 2021-01-06 at 2 16 01 PM" src="https://user-images.githubusercontent.com/2408094/103736499-5d0ed100-502b-11eb-9834-fdc0f8c1679c.png">

**Comment by kevinb9n** [comment]
Yes, definitely off by default - I think that to start with there should be no pressure to make it pretty at all.

**Comment by chosta** [comment]
I promise to work on it as soon as I'm done with my backlog tasks.

**Comment by vsrisuknimit** [comment]
Important issue.
